scripts/docker/build_docker.py

```diff
@@ -1,471 +1,377 @@
-                }
-            elif (proj == "sv-pipeline"):
-                build_time_args = {
-                    "GATKSV_PIPELINE_V1_RESOURCES_IMAGE": Project_Build.INTERM_RESOURCE_IMG + ":" + self.project_arguments.image_tag,
-                    "SV_PIPELINE_BASE_IMAGE": "sv-pipeline-base:" + self.project_arguments.image_tag
-                }
-            elif (proj.startswith("sv-pipeline")):
-                build_time_args = {
-                    "SV_PIPELINE_BASE_R_IMAGE": "sv-pipeline-children-r:" + self.project_arguments.image_tag
-                }
-
-            build_context = self.working_dir + "/dockerfiles/" + proj
-
-            remote_docker_repos = []
-            if (self.project_arguments.gcr_project is not None):
-                remote_docker_repos.append(
-                    "us.gcr.io/" + self.project_arguments.gcr_project)
-
-            docker = Docker_Build(proj, self.project_arguments.image_tag, build_context,
-                                  remote_docker_repos)
-            docker.build(build_time_args)
-            self.successfully_built_images.append(docker)
+            dependencies = ProjectBuilder.dependencies[target_name]
+            build_time_args = {} if dependencies is None else {
+                arg: f"{image_name}:{self.project_arguments.image_tag}" for image_name, arg in dependencies.items()
+            }
+
+            ImageBuilder(target_name, self).build(build_time_args, self.project_arguments.no_force_rebuild)
             print(colored('#################################################', 'magenta'))
 
-        for succ in self.successfully_built_images:
-            succ.push(self.project_arguments.update_latest)
+    def push(self):
+        if self.project_arguments.no_push:
+            return
+        for image in self.built_images:
+            ImageBuilder(image, self).push()
 
-        print(colored('BUILD PROCESS SUCCESS!', 'red'))
+    def build_and_push(self):
+        print("Project args:")
+        pprint.pprint(vars(self.project_arguments))
+        print("")
+        possible_tmp_dir_path = self.go_to_workdir()
+
+        try:
+            # start docker daemon, if one hasn't been started yet
+            os.system("open --background -a Docker && while ! docker system info > /dev/null 2>&1; do sleep 1; done")
+            self.build()
+            self.push()
+            print(colored('BUILD PROCESS SUCCESS!', 'red'))
+        finally:
+            if not self.project_arguments.skip_cleanup:
+                self.cleanup(possible_tmp_dir_path)
 
     def cleanup(self, tmp_dir_path):
-
-        # clean dangling images (i.e. those "<none>" images)
-        clean_dangling_images = "docker images --format \"{{.ID}}\"  --filter \"dangling=true\" -q --no-trunc | xargs docker rmi"
-        os.system(clean_dangling_images)
+        # clean dangling images (i.e. those "<none>" images), stopped containers, etc
+        os.system("docker system prune -f")
```

Comment:
This is probably a good suggestion. I haven't really had much problem with regretting the actions of docker system prune, but it would in the long run make build_docker.py nicer to have it only clean up its own crud. However, I definitely want to be stopping containers and killing dangling images, because it can easily happen during development that you get stopped containers. It may take me awhile to figure out a filter command that gets all the stuff we want without grabbing anything else (if you have a suggested filter, that may be a good starting point).

Reply:
Agree, we could keep the filter for a later PR.